cpp/src/lists/combine/concatenate_rows.cu

```diff
@@ -15,22 +15,170 @@
+      return offsets[row_index + 1] - offsets[row_index];
+    });
+
+  thrust::reduce_by_key(rmm::exec_policy(stream),
+                        keys,
+                        keys + (input.num_rows() * input.num_columns()),
+                        values,
+                        thrust::make_discard_iterator(),
+                        offsets->mutable_view().begin<offset_type>());
+
+  // convert to offsets
+  thrust::exclusive_scan(rmm::exec_policy(stream),
+                         offsets->view().begin<offset_type>(),
+                         offsets->view().begin<offset_type>() + input.num_rows() + 1,
+                         offsets->mutable_view().begin<offset_type>(),
+                         0);
+
+  // generate appropriate null mask
+  auto [null_mask, null_count] = [&]() {
+    // if the input doesn't contain nulls, no work to do
+    if (!build_null_mask) {
+      return std::pair<rmm::device_buffer, size_type>{rmm::device_buffer{}, 0};
+    }
+
+    // row is null if -all- input rows are null
+    if (null_policy == concatenate_null_policy::IGNORE) {
+      return cudf::detail::valid_if(
+        row_null_counts.begin(),
+        row_null_counts.begin() + input.num_rows(),
+        [num_columns = input.num_columns()] __device__(size_type null_count) {
+          return null_count != num_columns;
+        },
+        stream,
+        mr);
+    }
+
+    // row is null if -any- input rows are null
+    return cudf::detail::valid_if(
+      row_null_counts.begin(),
+      row_null_counts.begin() + input.num_rows(),
+      [] __device__(size_type null_count) { return null_count == 0; },
+      stream,
+      mr);
+  }();
+
+  return {std::move(offsets), std::move(null_mask), null_count};
+}
+
+rmm::device_uvector<size_type> generate_null_counts(table_device_view const& input,
+                                                    rmm::cuda_stream_view stream)
+{
+  rmm::device_uvector<size_type> null_counts(input.num_rows(), stream);
+
+  auto keys = thrust::make_transform_iterator(thrust::make_counting_iterator(size_t{0}),
+                                              [num_columns = input.num_columns()] __device__(
+                                                size_t i) -> size_type { return i / num_columns; });
+
+  auto null_values = thrust::make_transform_iterator(
+    thrust::make_counting_iterator(size_t{0}), [input] __device__(size_t i) -> size_type {
+      auto const col_index = i % input.num_columns();
+      auto const row_index = i / input.num_columns();
+      auto const& col      = input.column(col_index);
+      return col.null_mask() ? (bit_is_set(col.null_mask(), row_index + col.offset()) ? 0 : 1) : 0;
+    });
+
+  thrust::reduce_by_key(rmm::exec_policy(stream),
+                        keys,
+                        keys + (input.num_rows() * input.num_columns()),
+                        null_values,
+                        thrust::make_discard_iterator(),
+                        null_counts.data());
```

Comment:
Can we combine `generate_null_count` into `generate_regrouped_offsets_and_null_mask`? We have separate functions which in turn make separate kernel calls (like `reduce_by-key`) which seem to be expensive.

Reply:
No unfortunately.  The output of `generate_null_count()` is used in multiple places.  It is used to generate the null mask for the intermediate concatenated column as well as for generating the validity on the final column  (these two null masks are different).